From 2df2c35d41eb7c6eab5be7956a06340a8db558dd Mon Sep 17 00:00:00 2001
From: "H.J. Lu" <hjl.tools@gmail.com>
Date: Sun, 12 Jul 2015 12:06:38 -0700
Subject: [PATCH 2/2] Change x86 default to zlib-gabi

---
 gas/config/tc-i386.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gas/config/tc-i386.c b/gas/config/tc-i386.c
index 5a1e2b3..2bdb769 100644
--- a/gas/config/tc-i386.c
+++ b/gas/config/tc-i386.c
@@ -36,7 +36,7 @@
 #ifdef TE_LINUX
 /* Default to compress debug sections for Linux.  */
 enum compressed_debug_section_type flag_compress_debug
-  = COMPRESS_DEBUG_GNU_ZLIB;
+  = COMPRESS_DEBUG_GABI_ZLIB;
 #endif
 
 #ifndef REGISTER_WARNINGS
-- 
2.4.3

